### PR TITLE
Add support for editing directories

### PR DESCRIPTION
**Pull request by omus**
I wanted to be able to more easily edit packages. By adding directory support to `edit` you can now do things like `edit(Pkg.dir("TimeZones"))` to open an entire directory in your preferred editor.

The change required me to optionally add in arguments for line numbers. This was required by sublime to work correctly and I think it makes sense not to include them for other editors as well. I'll note that editors such as nano and TextEdit don't support opening directories but fail gracefully.

### CODE REVIEW

**Comment by omus** [comment]
I'll be updating the documentation

File: base/interactiveutil.jl [hunk 1]
```diff
@@ -23,31 +23,35 @@ function editor()
...
     if issrc
-        f = find_source_file(file)
-        f !== nothing && (file = f)
+        f = find_source_file(path)
+        f !== nothing && (path = f)
     end
     const no_line_msg = "Unknown editor: no line number information passed.\nThe method is defined at line $line."
     if startswith(name, "emacs") || name == "gedit"
-        spawn(`$command +$line $file`)
+        args = isfile(path) ? `+$line $path` : `$path`
```
Comment (stevengj) [inline base/interactiveutil.jl]: I guess this works, but it seems cleaner to me to use an array here for `args`, i.e. `["+$line" path] : [path]`.

**Comment by mbauman** [comment]
Huh, I've been doing exactly that for a long time.  I suppose the editors I've used (Atom and TextMate) just ignore line numbers for directories.  Would it be cleaner to only pass line numbers when they are given?  E.g., use a sentinel value for the default line number instead of checking isfile or isdir?

**Comment by omus** [comment]
Interesting. Maybe this is only really an issue for Sublime. Running `subl .:1` causes sublime to make a new file that matches the name of the directory you are in. If you do `subl .` it will open the directory.

**Comment by omus** [comment]
I switched the `isfile` checks to `line != 0`. I don't think any editors have line numbers starting at 0. Additionally the code no longer displays the line number warning when you aren't using them.

**Comment by omus** [comment]
I've updated the code to be up to date and did some refactoring to assist in future maintainability. @mbauman let me know what you think.

**Comment by stevengj** [comment]
LGTM.

**Comment by mbauman** [comment]
Agreed! This looks great.  Thanks @omus.

File: base/docs/helpdb.jl [hunk 2]
```diff
@@ -1335,11 +1335,11 @@ Get the value type of an associative collection type. Behaves similarly to `elty
...
 
 doc"""
-    edit(file::AbstractString, [line])
+    edit(path::AbstractString, [line])
 
-Edit a file optionally providing a line number to edit at. Returns to the julia prompt when you quit the editor.
+Edit a file or directory optionally providing a line number to edit the file at. Returns to the julia prompt when you quit the editor.
 """
-edit(file::AbstractString, line=?)
+edit(path::AbstractString, line=?)
```
Comment (swt30) [inline base/docs/helpdb.jl]: What does the question mark do in this context? I see it's in a few other places in helpdb but I've never seen this syntax before, and it's difficult to search for...

Comment (omus) [inline base/docs/helpdb.jl]: I believe it is just indicating that the line field is optional. http://julia.readthedocs.org/en/latest/stdlib/base/#Base.edit

**Pull request merged**